src/adapter/tracer/ptrace.rs

```diff
@@ -0,0 +1,278 @@
+/// for one query. When switching between queries, ensure you first call finish(),
+/// otherwise the new operations will be traced with the old.
+#[derive(Debug, Clone)]
+pub struct TracingAdapter<'vertex, AdapterT>
+where
+    AdapterT: Adapter<'vertex>,
+    AdapterT::Vertex: Debug + Clone + 'vertex,
+{
+    pub tracer: Rc<RefCell<Tracer>>,
+    inner: AdapterT,
+    _phantom: PhantomData<&'vertex ()>,
+}
+
+impl<'vertex, AdapterT> TracingAdapter<'vertex, AdapterT>
+where
+    AdapterT: Adapter<'vertex>,
+    AdapterT::Vertex: Debug + Clone + 'vertex,
+{
+    pub fn new(adapter: AdapterT, tracer: Rc<RefCell<Tracer>>) -> Self {
+        Self {
+            tracer,
+            inner: adapter,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn finish(self) -> Tracer {
+        // Ensure nothing is reading the trace i.e. we can safely stop interpreting.
+        let trace_ref = self.tracer.borrow_mut();
+        let new_trace = Tracer::new();
+        drop(trace_ref);
+        self.tracer.replace(new_trace)
+    }
+}
+
+impl<'vertex, AdapterT> Adapter<'vertex> for TracingAdapter<'vertex, AdapterT>
+where
+    AdapterT: Adapter<'vertex> + 'vertex,
+    AdapterT::Vertex: Debug + Clone + 'vertex,
+{
+    type Vertex = AdapterT::Vertex;
+
+    fn resolve_starting_vertices(
+        &self,
+        edge_name: &Arc<str>,
+        parameters: &EdgeParameters,
+        resolve_info: &ResolveInfo,
+    ) -> VertexIterator<'vertex, Self::Vertex> {
+        // We don't care about the first resolution since it's normally quick
+        // and it only occurs once.
+        self.inner
+            .resolve_starting_vertices(edge_name, parameters, resolve_info)
+    }
+
+    fn resolve_property<V: AsVertex<Self::Vertex> + 'vertex>(
+        &self,
+        contexts: ContextIterator<'vertex, V>,
+        type_name: &Arc<str>,
+        property_name: &Arc<str>,
+        resolve_info: &ResolveInfo,
+    ) -> ContextOutcomeIterator<'vertex, V, FieldValue> {
+        // For each resolution we want to know:
+        // 1. What are we resolving?
+        // 2. How long did the resolution take?
+        // 3. How many times did we resolve the same property?
+        //
+        // We are not collecting:
+        // 1. Number of times a function is called.
+        // 2. Whether or not any results were returned.
+
+        let call_id = FunctionCall::ResolveProperty(
+            resolve_info.vid(),
+            type_name.clone(),
+            property_name.clone(),
+        );
+
+        let inner_iter =
+            self.inner
+                .resolve_property(contexts, type_name, property_name, resolve_info);
+
+        let tracer_ref = self.tracer.clone();
+
+        Box::new(make_iter_with_perf_span(
+            inner_iter,
+            move |(context, value), duration| {
+                tracer_ref
+                    .borrow_mut()
+                    .record_time(call_id.clone(), duration);
```

Comment:
I'm not sure if this `.clone()` (+ similar ones elsewhere) to be able to use the `.entry()` API actually pays off.

On the very first `.record_time()` for a given function, using the `.entry()` API will prevent a double-lookup. Nice!

But the map of entries is quite small, and `.record_time()` is called way more than once. On every subsequent call, it's 1-2 `Arc::clone()` invocations which appear unnecessary because the key will be present.